### PR TITLE
Filter environment variables

### PR DESCRIPTION
**Pull request by topimiettinen**
Save all environment variables for later use in sandboxes, clear
environment and re-apply only whitelisted variables for the main
firejail process. All variables will be reapplied for the sandboxes by
env_apply().

### CODE REVIEW

**Comment by topimiettinen** [comment]
The commit eliminates the difference between start_application(0, fp) and start_application(1, fp) as now variables are applied unconditionally (because some of them don't exist anymore after clearenv()). To be honest, I'm not sure if there is now a change in behavior in some cases.

**Comment by topimiettinen** [comment]
Actually I'm pretty sure that all variables changed with setenv() in main Firejail process may be overwritten with the original values. So those setenv()s have to be changed to env_store()s.

File: src/firejail/env.c [hunk 1]
```diff
@@ -225,3 +226,46 @@ void env_apply(void) {
...
+	"FIREJAIL_FILE_COPY_LIMIT",
+	"FIREJAIL_PROMPT",
+	"FIREJAIL_QUIET",
+	"FIREJAIL_X11",
+	"LANG",
+	"LANGUAGE",
+	"LC_MESSAGES",
+	"PATH",
+	"SHELL",
+	"TMP",
```
Comment (Vincent43) [inline src/firejail/env.c:241]: Why `TMP` is whitelisted?

Comment (smitsohu) [reply src/firejail/env.c:241]: We used `TMP` to fix #2685

**Comment by smitsohu** [comment]
> To be honest, I'm not sure if there is now a change in behavior in some cases.

A while back I had what I think is the same problem with sanitizing the umask. So if you like you can  look out for `umask(orig_umask)` calls.

**Comment by topimiettinen** [comment]
In this version, the variables which Firejail uses internally are also moved to env storage for consistency. Thus the whitelist only applies to variables used by external libraries (libc/pthreads, libapparmor) and then it can be much shorter: only LANG, LANGUAGE, LC_MESSAGES.

Perhaps the sandboxed helper apps (fseccomp & friends) should also be run with a short list of variables: LANG etc. and internal variables starting with FIREJAIL_, otherwise they might need similar sanitizing (or perhaps they also need some sanitizing in any case to be sure). Only the final user app should get all variables restored.

**Comment by topimiettinen** [comment]
Latest build failure happens because tests expect that using `--env=FIREJAIL_TEST_ARGUMENTS=yes` (or exporting the environment variable FIREJAIL_TEST_ARGUMENTS from shell), the variable is passed also to the helper apps. Maybe this should be instead flagged with some other means, for example adding arguments to `faudit` like `faudit --test-arguments`. It would be simpler to just let the the variable pass.

**Comment by msva** [comment]
@topimiettinen ping?

Any progress here? ;)

We're all struggling (from env length checks) without your work ;)

**Comment by topimiettinen** [comment]
Sorry, I actually made a version which lets FIREJAIL_TEST_ARGUMENTS pass, but forgot to follow up. Thanks for the reminder. Let's see what the CIs think of the rebased version.

**Comment by topimiettinen** [comment]
Adding also PATH let the CI pass!

**Comment by msva** [comment]
@netblue30 ping? :D

**Comment by netblue30** [comment]
all merged!

**Pull request merged**